File: python/beeai_framework/tools/think.py
# Copyright 2025 © BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

from collections.abc import Callable

from pydantic import BaseModel, Field

from beeai_framework.context import RunContext
from beeai_framework.emitter import Emitter
from beeai_framework.tools import StringToolOutput, Tool, ToolRunOptions


class ThinkSchema(BaseModel):
    thoughts: str = Field(..., description="Precisely describe what you are thinking about.")
    next_step: list[str] = Field(..., description="Describe the tool you would need to use next and why.", min_length=1)


class ThinkTool(Tool[ThinkSchema]):
    name = "think"
    description = "Use when you want to think through a problem, clarify your assumptions, or break down complex steps before acting or responding."  # noqa: E501

    def __init__(self, *, extra_instructions: str = "", tool_output: str | Callable[[ThinkSchema], str] = "OK") -> None:
        super().__init__()
        self._tool_output = tool_output
        if extra_instructions:
            self.description += f" {extra_instructions}"

    @property
    def input_schema(self) -> type[ThinkSchema]:
        return ThinkSchema

    async def _run(self, input: ThinkSchema, options: ToolRunOptions | None, context: RunContext) -> StringToolOutput:
        output: str = self._tool_output(input) if isinstance(self._tool_output, Callable) else self._tool_output  # type: ignore
        return StringToolOutput(output)

    def _create_emitter(self) -> Emitter:
        return Emitter.root().child(
            namespace=["tool", "think"],
            creator=self,
        )

    async def clone(self) -> "ThinkTool":
        tool = ThinkTool(extra_instructions=self.description, tool_output=self._tool_output)
        tool.name = self.name
        tool.description = self.description
        tool._cache = await self.cache.clone()
        tool.middlewares = self.middlewares.copy()
        return tool

File: typescript/src/tools/think.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { z, ZodSchema } from "zod";
import { Tool, StringToolOutput, BaseToolRunOptions, BaseToolOptions } from "@/tools/base.js";
import { RunContext } from "@/context.js";
import { Emitter } from "@/emitter/emitter.js";

const ThinkSchema = z.object({
  thoughts: z.string().describe("Precisely describe what you are thinking about."),
});

type ThinkInput = z.infer<typeof ThinkSchema>;

export interface ThinkToolOptions extends BaseToolOptions<StringToolOutput> {
  extraInstructions?: string;
  toolOutput?: string | ((input: ThinkInput) => string);
  schema?: ZodSchema;
}

export class ThinkTool extends Tool<StringToolOutput, ThinkToolOptions> {
  public name = "think";
  public description =
    "Use when you want to think through a problem, clarify your assumptions, or break down complex steps before acting or responding.";

  public readonly emitter: Emitter<any>;

  protected _inputSchema: ZodSchema;
  protected _toolOutput: string | ((input: ThinkInput) => string);
  protected _extraInstructions: string;

  constructor(options: ThinkToolOptions = {}) {
    super(options);
    this._inputSchema = options.schema ?? ThinkSchema;
    this._toolOutput = options.toolOutput ?? "OK";
    this._extraInstructions = options.extraInstructions ?? "";

    if (this._extraInstructions) {
      this.description += ` ${this._extraInstructions}`;
    }

    this.emitter = this.createEmitter();
  }

  inputSchema(): ZodSchema {
    return this._inputSchema;
  }

  async _run(
    input: ThinkInput,
    _options: BaseToolRunOptions,
    _context: RunContext<typeof this>,
  ): Promise<StringToolOutput> {
    const output =
      typeof this._toolOutput === "function" ? this._toolOutput(input) : this._toolOutput;
    return new StringToolOutput(output);
  }

  protected createEmitter(): Emitter<any> {
    return Emitter.root.child({
      namespace: ["tool", "think"],
      creator: this,
    });
  }
}
